bin/picca_Pk1D.py

```diff
@@ -7,152 +7,367 @@
+        for part_index in range(num_parts):
+            # rebin exposures_diff spectrum
+            if (args.noise_estimate == 'rebin_diff'
+                    or args.noise_estimate == 'mean_rebin_diff'):
+                if linear_binning:
+                    exposures_diff_array[part_index] = rebin_diff_noise(
+                        delta_lambda, lambda_array[part_index],
+                        exposures_diff_array[part_index])
+                else:
+                    exposures_diff_array[part_index] = rebin_diff_noise(
+                        delta_log_lambda, log_lambda_array[part_index],
+                        exposures_diff_array[part_index])
+
+            # Fill masked pixels with 0.
+            if linear_binning:
+                #the resolution matrix does not need to have pixels filled in any way...
+                (lambda_new, delta_new, exposures_diff_new, ivar_new,
+                 num_masked_pixels) = fill_masked_pixels(
+                     delta_lambda, lambda_array[part_index],
+                     delta_array[part_index], exposures_diff_array[part_index],
+                     ivar_array[part_index], args.no_apply_filling)
+            else:
+                (log_lambda_new, delta_new, exposures_diff_new, ivar_new,
+                 num_masked_pixels) = fill_masked_pixels(
+                     delta_log_lambda, log_lambda_array[part_index],
+                     delta_array[part_index], exposures_diff_array[part_index],
+                     ivar_array[part_index], args.no_apply_filling)
+            if num_masked_pixels > args.nb_pixel_masked_max:
+                continue
+
+            # Compute pk_raw
+            if linear_binning:
+                k, pk_raw = compute_pk_raw(delta_lambda,
+                                           delta_new,
+                                           linear_binning=True)
+            else:
+                k, pk_raw = compute_pk_raw(delta_log_lambda,
+                                           delta_new,
+                                           linear_binning=False)
+
+            # Compute pk_noise
+            run_noise = False
+            if args.noise_estimate == 'pipeline':
+                run_noise = True
+            if linear_binning:
+                pk_noise, pk_diff = compute_pk_noise(delta_lambda,
+                                                     ivar_new,
+                                                     exposures_diff_new,
+                                                     run_noise,
+                                                     linear_binning=True)
+            else:
+                pk_noise, pk_diff = compute_pk_noise(delta_log_lambda,
+                                                     ivar_new,
+                                                     exposures_diff_new,
+                                                     run_noise,
+                                                     linear_binning=False)
```

Comment:
There are a lot of if conditions where only the linear binning changes in this function. Maybe for all of these, you can pass the linear_binning as an argument?

Reply:
it is an argument for the function already, mostly because things are converted from log-lambda binning to velocity binning internally if the binning is log-lambda. This is not done in case of linear binning.
But also the delta_lambda or delta_log_lambda needs to be put in here depending on binning. This is because here the function needs uniform sampling in whatever binning is used.

Alternatively one could also use the same variables for lambda/log_lambda and delta_lambda/delta_log_lambda and just pass that, but then it's less clear if a variable is in one binning or the other; or one could always pass lambda (or always log_lambda) and then convert the units internally, but that just shifts pain somewhere else...

In the long run I'd guess we need to make the whole computation more modular and properly saving properties like the binning, splitting, filling of pixels, resolution matrix etc in the class doing the computation, but that requires more time